### PR TITLE
Video streaming operators python bindings unit tests

### PR DESCRIPTION
**Pull request by cdinea**
# Python Binding Tests for Video Streaming Operators

## Overview

This PR adds Python binding tests (pytest) for the video streaming operators, complementing the existing C++ unit tests. The test suite validates the Python bindings for `VideoStreamingClientOp`, `StreamingServerResource`, `StreamingServerUpstreamOp`, and `StreamingServerDownstreamOp`.

### 📦 New Test Files

**Client Tests** (`operators/video_streaming/video_streaming_client/python/tests/`):
- `test_streaming_client_op_bindings.py` - 30 test functions (38 test executions with parametrized variants)
- `conftest.py` - Pytest fixtures for client tests
- `README.md` - Detailed test documentation

**Server Tests** (`operators/video_streaming/video_streaming_server/python/tests/`):
- `test_streaming_server_ops_bindings.py` - 42 test functions (51 test executions with parametrized variants)
- `conftest.py` - Pytest fixtures for server tests
- `README.md` - Detailed test documentation

**Summary Documentation**:
- `PYTEST_SUMMARY.md` - High-level overview of all Python binding tests
- `UNIT_TESTS_SUMMARY.md` - Updated with cross-references to pytest documentation

### 🧪 Test Coverage (89 Test Executions from 72 Test Functions)

> **Note on Test Counts:** The test suite contains 72 test functions. When pytest runs parametrized tests (tests with multiple input variants), each variant counts as a separate test execution. This results in 89 total test executions: 72 base test functions + 17 parametrized variants (8 client + 9 server).

#### Client Tests (38 test executions)
- **Operator Binding Tests** (20+ tests)
  - Basic creation and configuration
  - Parameter validation (video dimensions, network settings, streaming modes)
  - Method availability and inheritance
  - Memory management and operator reuse
  - Edge cases (extreme resolutions, FPS, ports)
  
- **Integration Tests** (2 tests)
  - Application context integration
  - Fragment-based workflow
  
- **Mock Data Tests** (6 tests)
  - CuPy and NumPy array handling
  - Various resolutions and color formats
  - Grayscale and float frame data
  - Seeded random data validation
  
- **Compute Method Tests** (5 tests)
  - Method accessibility and signature
  - Mock frame processing
  - Various resolutions and data types

#### Server Tests (51 test executions)
- **Resource Binding Tests** (23 test executions)
  - Resource creation and configuration
  - Parameter validation (video parameters with 4 variants, ports with 4 variants, directions with 4 variants)
  - Bidirectional streaming setup
  - Memory management and isolation
  
- **Upstream Operator Tests** (6 tests)
  - Operator creation with resources
  - Method availability and inheritance
  - Shared resource handling
  
- **Downstream Operator Tests** (6 tests)
  - Operator creation with resources
  - Method availability and inheritance
  - Shared resource handling
  
- **Integration Tests** (4 tests)
  - Bidirectional server setup
  - Multiple servers on different ports
  - Application and fragment context
  - Resource isolation
  
- **Mock Data Tests** (7 tests)
  - Resource with various frame dimensions
  - Upstream/downstream with mock frames
  - Bidirectional streaming with mock data
  - NumPy and CuPy compatibility
  
- **Compute Method Tests** (5 tests)
  - Upstream compute() validation
  - Downstream compute() validation
  - Bidirectional compute flow

### 🔧 Infrastructure Changes

1. **CMake Integration**
   - Added `add_python_tests()` calls in both client and server `CMakeLists.txt`
   - Tests are registered with CTest for `./holohub test` integration
   
2. **Docker Support**
   - Added pytest and pytest-cov to `applications/video_streaming/Dockerfile`
   
3. **Pytest Fixtures**
   - Local conftest files with common fixtures (`app`, `fragment`, `mock_image`, `op_input_factory`, `op_output`, `execution_context`)
   - Handles `--confcutdir` isolation from `add_python_tests()`
   - Fixed bug in `mock_image` fixture for float dtype handling

4. **Documentation**
   - Comprehensive test coverage documentation
   - Clear cross-references between all test READMEs
   - CUDA compatibility warnings
   - Troubleshooting guides

### Run All Python Binding Tests

```bash
# Run both client and server pytest tests
./holohub test video_streaming --cuda 12 --ctest-options="-R streaming.*pytest -VV"
```

### Run Client Tests Only

```bash
./holohub test video_streaming --cuda 12 --ctest-options="-R video_streaming_client_pytest -VV"
```

### Run Server Tests Only

```bash
./holohub test video_streaming --cuda 12 --ctest-options="-R video_streaming_server_pytest -VV"
```

### Run Specific Test Class

```bash
# Example: Run only compute method tests
./holohub test video_streaming --cuda 12 --ctest-options="-R 'Compute' -VV"
```

### Direct Pytest Execution (for development)

```bash
# From within the Docker container or with proper environment setup
cd operators/video_streaming/video_streaming_client/python
pytest tests/ -v

cd operators/video_streaming/video_streaming_server/python
pytest tests/ -v
```

### ✅ Success Criteria

**Test Execution:**

100% tests passed, 0 tests failed out of 89
Total Test time (real) = ~20 sec

**Breakdown:**
- ✅ 38/38 client test executions passing (30 functions + 8 parametrized variants)
- ✅ 51/51 server test executions passing (42 functions + 9 parametrized variants)

**Sample Output:**

`The following tests passed:
video_streaming_client_pytest.test_streaming_client_op_bindings.py::TestStreamingClientOpBinding::test_operator_creation_basic
video_streaming_client_pytest.test_streaming_client_op_bindings.py::TestStreamingClientOpBinding::test_operator_name
video_streaming_client_pytest.test_streaming_client_op_bindings.py::TestStreamingClientOpBinding::test_video_parameters[640-480-30]
video_streaming_client_pytest.test_streaming_client_op_bindings.py::TestStreamingClientOpBinding::test_video_parameters[1280-720-60]
...
video_streaming_server_pytest.test_streaming_server_ops_bindings.py::TestStreamingServerResourceBinding::test_resource_creation_basic
video_streaming_server_pytest.test_streaming_server_ops_bindings.py::TestStreamingServerUpstreamOpBinding::test_operator_creation_basic
...
video_streaming_server_pytest.test_streaming_server_ops_bindings.py::TestBidirectionalServerCompute::test_bidirectional_compute_flow
100% tests passed, 0 tests failed out of 89
`

### ✅ Functional Requirements

1. **All 89 test executions pass** (72 test functions + 17 parametrized variants) on CUDA 12 systems
2. **Tests are discoverable** via `./holohub test video_streaming`
3. **Tests run in Docker** with all dependencies available
4. **Tests validate**:
   - Python bindings are correctly exposed
   - Parameters can be set and retrieved
   - Operators can be instantiated in Application/Fragment contexts
   - `compute()` methods are accessible and functional
   - Mock data flows through operators correctly
   - Memory management works (no leaks or crashes)

### ✅ Code Quality Requirements

1. **Test Organization**:
   - Tests grouped by functionality (binding, integration, mock data, compute)
   - Clear test names describing what is being validated
   - Comprehensive docstrings
   - Parametrized tests for multiple scenarios

2. **Fixtures**:
   - Reusable fixtures in conftest.py
   - Factory patterns for flexible test setup
   - Proper isolation between tests
   - Fixed float dtype handling bug

3. **Documentation**:
   - Individual README files for client and server tests
   - High-level PYTEST_SUMMARY.md documenting all tests
   - UNIT_TESTS_SUMMARY.md updated with pytest references
   - Clear cross-references between all documentation
   - CUDA 12/13 compatibility warnings
   - Troubleshooting guides

### ✅ Integration Requirements

1. **CMake Integration**:
   - Tests registered with CTest
   - Proper test naming convention (`video_streaming_*_pytest.*`)
   - Tests run via `./holohub test` command

2. **CI/CD Ready**:
   - Tests run in Docker container
   - No manual setup required
   - Deterministic results
   - All links in documentation verified

3. **Compatibility**:
   - Works with CUDA 12 (required for client libraries)
   - Compatible with Holoscan SDK v3.8.0
   - No conflicts with existing C++ unit tests

## CUDA Version Compatibility Note

⚠️ **Important:** The video streaming client tests require **CUDA 12**. If you're using CUDA 13, you must specify `--cuda 12` when running the `./holohub` script.

**For CUDA 13 systems:**
```bash
./holohub build video_streaming --cuda 12 --configure-args='-DBUILD_TESTING=ON'
./holohub test video_streaming --cuda 12 --ctest-options="-R streaming.*pytest -VV"
```

**Why CUDA 12?** The video streaming client libraries are currently built against CUDA 12 runtime libraries. Running tests on CUDA 13 without the `--cuda 12` flag will result in test failures due to library incompatibilities.

## Testing Checklist

- [x] All 89 test executions pass locally
- [x] Tests run successfully via `./holohub test`
- [x] Tests run in Docker container
- [x] CMake properly registers tests with CTest
- [x] Documentation updated (READMEs, test summaries)
- [x] CUDA 12/13 compatibility documented
- [x] No regressions in existing C++ unit tests
- [x] Code follows holohub testing patterns (similar to pixelator operator)
- [x] All documentation links verified (link checker passes)
- [x] Float dtype bug in mock_image fixture fixed

## Key Bug Fixes

1. **Float dtype handling in mock_image fixture**: Fixed `TypeError` where `rng.uniform()` was incorrectly passed a `dtype` parameter. Now generates float arrays and casts to requested dtype if needed.
2. **Documentation links**: Corrected relative paths to `PYTEST_SUMMARY.md` in client and server test READMEs.
3. **Test counts**: Clarified distinction between test functions (72) and test executions (89) throughout documentation.

## Review Notes

This PR focuses exclusively on **Python binding validation** - it does not test the full network I/O functionality (that's covered by integration tests). The tests verify:
- Bindings are correctly exposed to Python
- Parameters work as expected
- Operators can be created and configured
- `compute()` methods are accessible (functional validation with mocks)
- Memory management across Python/C++ boundary

> **Note on Fixtures:** The `add_python_tests()` CMake function uses `--confcutdir` to isolate tests, which prevents access to the root `conftest.py`. Common fixtures are duplicated in local conftest files to maintain CTest integration while providing necessary test utilities.

## Related Documentation

- **PYTEST_SUMMARY.md** - High-level overview of all Python binding tests

## Summary by CodeRabbit

* **Tests**
  * Added comprehensive Python unit test suites for video streaming client and server bindings covering creation, configuration, compute behavior, mock-frame scenarios, integration flows, and CUDA variants.

* **Documentation**
  * Added detailed test READMEs and summaries with execution workflows, troubleshooting, test organization, and CUDA 12 compatibility guidance.

* **Chores**
  * Enabled Python test execution in builds, added pytest/pytest-cov to the test image, and adjusted build/runtime handling to prefer system CUDA and use symlinked runtime libraries.

### CODE REVIEW

**Comment by cdinea** [comment]
@coderabbitai full review

**Comment by cdinea** [comment]
@coderabbitai full review

File: operators/video_streaming/PYTEST_SUMMARY.md [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+grep "failed" test_results.log
+```
+
+## ✅ Acceptance Criteria
+
+All tests must pass with the following criteria:
+
+| Criteria | Expected | Status |
+|----------|----------|--------|
+| **Total Tests** | 61 tests discovered and run | ✅ PASS |
```
[the remark this answers is not in the document]
Comment (cdinea) [reply operators/video_streaming/PYTEST_SUMMARY.md]: @greptile-apps addressed this feedback in [this](https://github.com/nvidia-holoscan/holohub/pull/1225/commits/2f29ea982a4fdf3cdd3c016f6b0c0a9281b9a793) commit

File: operators/video_streaming/video_streaming_client/python/tests/conftest.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+            xp = np
+        else:
+            raise ValueError(f"Unknown backend: {backend}")
+        rng = xp.random.default_rng(seed)
+        dtype = xp.dtype(dtype)
+        if dtype.kind in "ui":
+            img = rng.integers(0, 256, size=shape, dtype=dtype, endpoint=False)
+        elif dtype.kind == "f":
+            img = rng.uniform(0.0, 1.0, size=shape, dtype=dtype)
+        else:
```
[the remark this answers is not in the document]
Comment (cdinea) [reply operators/video_streaming/video_streaming_client/python/tests/conftest.py:136]: addressed this feedback in [this](https://github.com/nvidia-holoscan/holohub/pull/1225/commits/ff6d23507b48037edb5fbcc09ff434504437bc18) commit

File: operators/video_streaming/video_streaming_server/python/tests/conftest.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+            xp = np
+        else:
+            raise ValueError(f"Unknown backend: {backend}")
+        rng = xp.random.default_rng(seed)
+        dtype = xp.dtype(dtype)
+        if dtype.kind in "ui":
+            img = rng.integers(0, 256, size=shape, dtype=dtype, endpoint=False)
+        elif dtype.kind == "f":
+            img = rng.uniform(0.0, 1.0, size=shape, dtype=dtype)
+        else:
```
[the remark this answers is not in the document]
Comment (cdinea) [reply operators/video_streaming/video_streaming_server/python/tests/conftest.py:166]: feedback already addressed

File: operators/video_streaming/PYTEST_SUMMARY.md [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+| **Client Tests** | 30 tests | ✅ PASS (30/30) |
+| **Server Tests** | 31 tests (18 Resource + 6 Upstream + 6 Downstream + 4 Integration) | ✅ PASS (31/31) |
+| **Pass Rate** | 100% of tests passing | ✅ PASS (61/61 = 100%) |
```
[the remark this answers is not in the document]
Comment (cdinea) [reply operators/video_streaming/PYTEST_SUMMARY.md]: feedback already addressed

[the remark this answers is not in the document]
Comment (cdinea) [reply operators/video_streaming/PYTEST_SUMMARY.md]: feedback already addressed

**Comment by bhashemian** [comment]
@coderabbitai full review

**Comment by cdinea** [comment]
> Thanks @cdinea for contributing the Python unit tests. They look good to me, and I’ve tested them with HSDK versions 3.5 and 3.7, and they work flawlessly. However, the client encountered an issue with version 3.8. Could you please investigate this? Thanks

Thank you for the review @bhashemian - I will check the client app with the newly released HSDK 3.8 and report back - I believe in my experiments I was trying with an unreleased version of 3.8 or an rc.

**Comment by cdinea** [comment]
@bhashemian I have modified the CMake because the NGC-bundled libraries included CUDA 12.0.107, but Holoscan SDK 3.8.0 requires CUDA 12.8+ which has the cudaGraphConditionalHandleCreate symbol.As a soluution I replace bundled CUDA 12.0.107 with symlinks to system CUDA 12.8+. In the long run we can go through NGC approval for a wrapper binary without CUDA.I also udpated RPATH to prioritize system libraries

all unit tests pass on a clean env on my machine
`
The following tests passed:
        video_streaming_client_op_unit_tests
        video_streaming_server_ops_unit_tests

100% tests passed, 0 tests failed out of 2

Label Time Summary:
unit                      =   0.12 sec*proc (2 tests)
video_streaming_client    =   0.06 sec*proc (1 test)
video_streaming_server    =   0.06 sec*proc (1 test)

Total Test time (real) =   0.13 sec`
